### ISSUE TITLE
cyradm warnings with perl 5.16 (centos7)

### ISSUE THREAD

**Issue by Isma399** [cyrusimap/cyrus-imapd#2747]
Hello,

I've got this error after connecting to cyrus with cyradm (as root or cyrus user):

```
# cyradm -u cyrus localhost
Variable "$cyrref" will not stay shared at /usr/lib64/perl5/vendor_perl/Cyrus/IMAP/Shell.pm line 724.
Variable "$lfh" will not stay shared at /usr/lib64/perl5/vendor_perl/Cyrus/IMAP/Shell.pm line 726.
Duplicate specification "server|s=s" for option "s"
```

I can make operation on mailbox (lam, sam, xfer, ..), everything seems to work fine but I'm not confident to put that in production..

Cyrus version is 3.08 installed with rpm on Centos7.

I've build the rpms, so maybe I've made mistake at this step.

cyrus was build like that :

```
# cyr_buildinfo
{
  "component": {
    "event_notification": true,
    "gssapi": true,
    "autocreate": true,
    "idled": true,
    "httpd": true,
    "kerberos_v4": false,
    "murder": true,
    "nntpd": true,
    "replication": true,
    "sieve": true,
    "calalarmd": true,
    "objectstore": false,
    "backup": true
  },
  "dependency": {
    "ldap": true,
    "openssl": true,
    "pcre": true,
    "clamav": true
  },
  "database": {
    "mysql": false,
    "pgsql": false,
    "sqlite": true,
    "lmdb": false
  },
  "search": {
    "squat": true,
    "sphinx": false,
    "xapian": false,
    "xapian_flavor": "none"
  },
  "hardware": {
    "sse42": true
  }
}

```

```
# perl --version
This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi
(with 38 registered patches, see perl -V for more detail)
```
Although I'm quite not a perl monger (python is my usual language), I'll be glad to give some help. 

Thank you

**Comment by hagedose** [comment]
FWIW, I have a test system running RHEL 7 with the same Perl version, and I don't see that message. I also built the RPM myself, so I guess there is some difference in our build process.

```
cyr_buildinfo
{
  "component": {
    "event_notification": true,
    "gssapi": true,
    "autocreate": false,
    "idled": true,
    "httpd": false,
    "kerberos_v4": false,
    "murder": true,
    "nntpd": false,
    "replication": true,
    "sieve": true,
    "calalarmd": false,
    "objectstore": false,
    "backup": true
  },
  "dependency": {
    "ldap": true,
    "openssl": true,
    "pcre": true,
    "clamav": false
  },
  "database": {
    "mysql": false,
    "pgsql": false,
    "sqlite": true,
    "lmdb": false
  },
  "search": {
    "squat": true,
    "sphinx": false,
    "xapian": true,
    "xapian_flavor": "vanilla"
  },
  "hardware": {
    "sse42": true
  }
}
```

**Comment by Isma399** [comment]
Thanks for your answer.
I have used a modified Fedora spec file, you can find it here:
https://github.com/Isma399/cyrus3.0.8-rpms/tree/master/SPECS

My goal was to use the autocreate patch.
Il also need murder, replication, sieve, idled components.

I've build packages with this command:
`rpmbuild -ba rpmbuild/SPECS/cyrus-imapd.spec`

If you have any time to look at this spec file, thanks!

**Comment by elliefm** [comment]
It looks like the warning is talking about `sub showacl()`, whose declaration is nested inside `sub _sc_listacl()` (despite named subs in perl having package scope, not block scope -- i.e. this sub is not as "hidden" as it seems like it might be). 

I believe the warnings are telling us that those two variables it grabs from the "outer" subroutine will only be shared on the first invocation of `showacl()`, and that they will preserve whatever value they had at that time for future invocations of `showacl()` (despite them maybe being changed by other calls to `_sc_listact()` in the meantime)

This all appears to be irrelevant because, to my reading, nothing seems to call `showacl()`?  (Not even `_sc_listacl()` calls it!)  Like, I think we could safely just delete that `sub showacl()` definition (highlighted lines here: https://github.com/cyrusimap/cyrus-imapd/blob/0bbead0889a140853474b87b29ad173101301c0c/perl/imap/IMAP/Shell.pm#L721-L736) and nothing would even notice?

There might be something tricksy going on here that I'm not yet understanding... but skimming through the history it _looks_ like maybe an artifact of a bad merge in 2003  (sigh) that just never caused any problems until perl started warning about bugs in the dead code(!)

**Comment by elliefm** [comment]
This will be fixed in 3.2.2
